File: src/main/java/org/mitallast/queue/crdt/routing/RoutingBucket.java
package org.mitallast.queue.crdt.routing;

import javaslang.collection.HashMap;
import javaslang.collection.Map;
import javaslang.collection.Seq;
import javaslang.collection.Set;
import javaslang.control.Option;
import org.mitallast.queue.common.codec.Codec;
import org.mitallast.queue.common.codec.Message;
import org.mitallast.queue.transport.DiscoveryNode;

public class RoutingBucket implements Message {
    public static final Codec<RoutingBucket> codec = Codec.Companion.of(
        RoutingBucket::new,
        RoutingBucket::index,
        RoutingBucket::replicaSeq,
        RoutingBucket::resourceSeq,
        Codec.Companion.intCodec(),
        Codec.Companion.seqCodec(RoutingReplica.codec),
        Codec.Companion.seqCodec(Resource.codec)
    );

    private final int index;
    private final Map<Long, RoutingReplica> replicas;
    private final Map<Long, Resource> resources;

    public RoutingBucket(int index) {
        this(index, HashMap.empty(), HashMap.empty());
    }

    public RoutingBucket(int index, Seq<RoutingReplica> replicas, Seq<Resource> resources) {
        this.index = index;
        this.replicas = replicas.toMap(RoutingReplica::id, r -> r);
        this.resources = resources.toMap(Resource::id, r -> r);
    }

    public RoutingBucket(int index, Map<Long, RoutingReplica> replicas, Map<Long, Resource> resources) {
        this.index = index;
        this.replicas = replicas;
        this.resources = resources;
    }

    public int index() {
        return index;
    }

    public Seq<RoutingReplica> replicaSeq() {
        return replicas.values();
    }

    public Map<Long, RoutingReplica> replicas() {
        return replicas;
    }

    public Option<RoutingReplica> replica(DiscoveryNode member) {
        return replicas.values().find(replica -> replica.member().equals(member));
    }

    public boolean exists(DiscoveryNode member) {
        return replicas.values().exists(replica -> replica.member().equals(member));
    }

    public Seq<Resource> resourceSeq() {
        return resources.values();
    }

    public Map<Long, Resource> resources() {
        return resources;
    }

    public RoutingBucket withResource(Resource resource) {
        return new RoutingBucket(
            index,
            replicas,
            resources.put(resource.id(), resource)
        );
    }

    public RoutingBucket withoutResource(long resource) {
        return new RoutingBucket(
            index,
            replicas,
            resources.remove(resource)
        );
    }

    public boolean hasResource(long id) {
        return resources.containsKey(id);
    }

    public Resource resource(long id) {
        return resources.getOrElse(id, null);
    }

    public RoutingBucket withReplica(RoutingReplica member) {
        return new RoutingBucket(
            index,
            replicas.put(member.id(), member),
            resources
        );
    }

    public RoutingBucket filterReplicas(Set<DiscoveryNode> members) {
        return new RoutingBucket(
            index,
            replicas.filterValues(replica -> members.contains(replica.member())),
            resources
        );
    }

    public RoutingBucket withoutReplica(long replica) {
        return new RoutingBucket(
            index,
            replicas.remove(replica),
            resources
        );
    }

    @Override
    public String toString() {
        return "RoutingBucket{" +
            "index=" + index +
            ", members=" + replicas +
            ", resources=" + resources +
            '}';
    }
}

File: src/main/java/org/mitallast/queue/crdt/routing/RoutingBucket.kt
package org.mitallast.queue.crdt.routing

import javaslang.collection.HashMap
import javaslang.collection.Map
import javaslang.collection.Seq
import javaslang.collection.Set
import javaslang.control.Option
import org.mitallast.queue.common.codec.Codec
import org.mitallast.queue.common.codec.Message
import org.mitallast.queue.transport.DiscoveryNode

class RoutingBucket : Message {
    val index: Int
    val replicas: Map<Long, RoutingReplica>
    val resources: Map<Long, Resource>

    constructor(index: Int, replicas: Seq<RoutingReplica>, resources: Seq<Resource>) {
        this.index = index
        this.replicas = replicas.toMap({ it.id }) { it }
        this.resources = resources.toMap({ it.id }) { it }
    }

    @JvmOverloads constructor(index: Int, replicas: Map<Long, RoutingReplica> = HashMap.empty(), resources: Map<Long, Resource> = HashMap.empty()) {
        this.index = index
        this.replicas = replicas
        this.resources = resources
    }

    fun replicaSeq(): Seq<RoutingReplica> {
        return replicas.values()
    }

    fun replica(member: DiscoveryNode): Option<RoutingReplica> {
        return replicas.values().find { replica -> replica.member == member }
    }

    fun exists(member: DiscoveryNode): Boolean {
        return replicas.values().exists { replica -> replica.member == member }
    }

    fun resourceSeq(): Seq<Resource> {
        return resources.values()
    }

    fun withResource(resource: Resource): RoutingBucket {
        return RoutingBucket(
            index,
            replicas,
            resources.put(resource.id, resource)
        )
    }

    fun withoutResource(resource: Long): RoutingBucket {
        return RoutingBucket(
            index,
            replicas,
            resources.remove(resource)
        )
    }

    fun hasResource(id: Long): Boolean {
        return resources.containsKey(id)
    }

    fun resource(id: Long): Resource {
        return resources.getOrElse(id, null)
    }

    fun withReplica(member: RoutingReplica): RoutingBucket {
        return RoutingBucket(
            index,
            replicas.put(member.id, member),
            resources
        )
    }

    fun filterReplicas(members: Set<DiscoveryNode>): RoutingBucket {
        return RoutingBucket(
            index,
            replicas.filterValues { replica -> members.contains(replica.member) },
            resources
        )
    }

    fun withoutReplica(replica: Long): RoutingBucket {
        return RoutingBucket(
            index,
            replicas.remove(replica),
            resources
        )
    }

    override fun toString(): String {
        return "RoutingBucket{" +
            "index=" + index +
            ", members=" + replicas +
            ", resources=" + resources +
            '}'
    }

    companion object {
        val codec = Codec.of(
            ::RoutingBucket,
            RoutingBucket::index,
            RoutingBucket::replicaSeq,
            RoutingBucket::resourceSeq,
            Codec.intCodec(),
            Codec.seqCodec(RoutingReplica.codec),
            Codec.seqCodec(Resource.codec)
        )
    }
}
